predicthq/kafka/consumer.py

```diff
@@ -0,0 +1,163 @@
+from .payload import unpack_kafka_payload
+
+
+log = log = logging.getLogger(__name__)
+
+
+Tp = namedtuple('Tp', ['topic', 'partition'])
+
+Message = namedtuple('Message', ['id', 'payload', 'ref'])
+
+
+def get_kafka_consumer(bootstrap_servers: List[str], group_id: str, kafka_custom_config: Dict[str, str]) -> confluent_kafka.Consumer:
+
+    configuration = KAFKA_CONSUMER_BASE_CONFIGURATION.copy()
+    if kafka_custom_config:
+        configuration.update(kafka_custom_config)
+    configuration.update({
+        'bootstrap.servers': ",".join(bootstrap_servers),
+        'group.id': group_id,
+    })
+    return confluent_kafka.Consumer(configuration, logger=log)
+
+
+def format_batch_ref(batch_ref):
+    return ','.join('{}[{}]'.format(topic, ','.join('{}={}:{}'.format(partition, min_offset, max_offset + 1)
+                                                    for partition, (min_offset, max_offset)
+                                                    in partition_dict.items()))
+                    for topic, partition_dict in batch_ref.items())
+
+
+def _group_messages(messages):
+    batch_ref = {}
+    for topic, m_topic in itertools.groupby(messages, lambda x: x.topic()):
+        for partition, m_partition in itertools.groupby(m_topic, lambda x: x.partition()):
+            tp = Tp(topic, partition)
+            if tp not in batch_ref:
+                batch_ref[tp] = []
+            batch_ref[tp] += list(m_partition)
+    return batch_ref
+
+
+class PhqKafkaConsumer(object):
+    def __init__(
+        self, input_topic: str, group_id: str, batch_size: int, consumer_timeout_ms: int,
+        kafka_bootstrap_servers: List[str], kafka_consumer_config: Dict[str, str] = None
+    ):
+        self.closed = False
+
+        self.input_topic = input_topic
+
+        self._consumer_timeout_ms = consumer_timeout_ms
+
+        self._consumer_group = group_id
+        self._consumer_batch_size = batch_size
+        self._consumer = get_kafka_consumer(
+            kafka_bootstrap_servers,
+            self._consumer_group,
+            kafka_consumer_config
+        )
+
+        self._consumer.subscribe([self.input_topic])
+
+    def _format_message_batch(self, kafka_messages):
+        messages = []
+        batch_ref = {}
+        for tp, batch in _group_messages(kafka_messages).items():
+            topic = tp.topic
+            partition = tp.partition
+
+            if topic == self.input_topic:
+                if topic not in batch_ref:
+                    batch_ref[topic] = {}
+
+                for message in batch:
+                    # Skip null messages - they are used to 'delete' keys from
+                    # compacted topics.
+                    if message.value() is None:
+                        continue
+
+                    if message.error():
+                        error = message.error()
+                        if error.code() == confluent_kafka.KafkaError._PARTITION_EOF:
```

Comment:
why we need to handle this?

Reply:
if we don't, and there is no more message in the queue (consumer hanging and waiting for message)

Reply:
In our case, it's not error, as we except or consumer to still wait for messages when reaching the end of the topic.